components/script/dom/document.rs

```diff
@@ -2016,6 +2024,118 @@ impl Document {
+        if !self.get_allow_fullscreen() {
+            error = true;
+        }
+        // TODO fullscreen is supported
+        // TODO This algorithm is allowed to request fullscreen.
+
+        // Step 5 Parallel start
+
+        let window = self.window();
+        // Step 6
+        if !error {
+            let event = ConstellationMsg::SetFullscreenState(true);
+            window.upcast::<GlobalScope>().constellation_chan().send(event).unwrap();
+        }
+
+        // Step 7
+        let trusted_pending = Trusted::new(pending);
+        let trusted_promise = TrustedPromise::new(promise.clone());
+        let handler = ElementPerformFullscreenEnter::new(trusted_pending, trusted_promise, error);
+        let script_msg = CommonScriptMsg::RunnableMsg(ScriptThreadEventCategory::EnterFullscreen, handler);
+        let msg = MainThreadScriptMsg::Common(script_msg);
+        window.main_thread_script_chan().send(msg).unwrap();
+
+        promise
+    }
+
+    // https://fullscreen.spec.whatwg.org/#exit-fullscreen
+    #[allow(unrooted_must_root)]
+    pub fn exit_fullscreen(&self) -> Rc<Promise> {
+        let global = self.global();
+        // Step 1
+        let promise = Promise::new(global.r());
+        // Step 2
+        if self.fullscreen_element.get().is_none() {
+            promise.reject_error(global.get_cx(), Error::Type(String::from("fullscreen is null")));
+            return promise
+        }
+        // TODO Step 3-6
+        let element = self.fullscreen_element.get().unwrap();
+
+        // Step 7 Parallel start
+
+        let window = self.window();
+        // Step 8
+        let event = ConstellationMsg::SetFullscreenState(false);
+        window.upcast::<GlobalScope>().constellation_chan().send(event).unwrap();
+
+        // Step 9
+        let trusted_element = Trusted::new(element.r());
+        let trusted_promise = TrustedPromise::new(promise.clone());
+        let handler = ElementPerformFullscreenExit::new(trusted_element, trusted_promise);
+        let script_msg = CommonScriptMsg::RunnableMsg(ScriptThreadEventCategory::ExitFullscreen, handler);
+        let msg = MainThreadScriptMsg::Common(script_msg);
+        window.main_thread_script_chan().send(msg).unwrap();
+
+        promise
+    }
+
+    pub fn set_fullscreen_element(&self, element: Option<&Element>) {
+        self.fullscreen_element.set(element);
+    }
+
+    pub fn set_allow_fullscreen(&self, value: bool) {
+        self.allow_fullscreen.set(value);
+    }
+
+    pub fn get_allow_fullscreen(&self) -> bool {
+        // https://html.spec.whatwg.org/multipage/#allowed-to-use
+        match self.browsing_context() {
+            // Step 1
+            None => true,
+            Some(_) => {
+                // Step 2
+                let window = self.window();
+                if window.is_top_level() {
+                    true
+                } else {
+                    // Step 3
+                    self.allow_fullscreen.get()
```

Comment:
For right now, we can use `window.GetFrameElement().map(|el el.has_attribute(atom!("allowfullscreen"))).unwrap_or(false)`. This will work correctly for same-origin iframes and fail for cross-origin ones.